src/libsystemd/sd-json/json-util.c

```diff
@@ -344,17 +345,25 @@ int json_dispatch_ifindex(const char *name, sd_json_variant *variant, sd_json_di
 int json_dispatch_log_level(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
         int *log_level = ASSERT_PTR(userdata), r, t;
 
+        /* If SD_JSON_STRICT is set, we'll refuse attempts to set the log level to null. If SD_JSON_RELAX is
+         * set we'll turn null (and any negative log level) into LOG_NULL (which when used as max log level
+         * means: no logging). Otherwise we turn null into LOG_INFO (which is typically our default). */
+
         if (sd_json_variant_is_null(variant)) {
-                *log_level = -1;
+                if (FLAGS_SET(flags, SD_JSON_STRICT))
+                        return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' may not be null.", strna(name));
+
+                *log_level = FLAGS_SET(flags, SD_JSON_RELAX) ? LOG_NULL : LOG_INFO;
```

Comment:
> rebased, and dropped the logic to allow clients to set the log level to null.

Looks like it wasn't dropped after all. `null` is accepted and resets the log level to INFO as far as I can see. It would be really weird if one day it started drop all the logs instead (which the comment says can happen with `SD_JSON_RELAX`).

Reply:
> null is accepted and resets the log level to INFO

Looks like it's what happens:
```sh
Jan 19 13:56:04 systemd-resolved[1139]: varlink: New incoming connection.
Jan 19 13:56:04 systemd-resolved[1139]: varlink: Connections of user 0: 0 (of 576 max)
Jan 19 13:56:04 systemd-resolved[1139]: varlink-24-24: Setting state idle-server
Jan 19 13:56:04 systemd-resolved[1139]: varlink-24-24: Received message: {"method":"io.systemd.service.SetLogLevel","parameters":{"level":null}}
Jan 19 13:56:04 systemd-resolved[1139]: varlink-24-24: Changing state idle-server → processing-method
Jan 19 13:56:04 systemd-resolved[1139]: Received io.systemd.service.SetLogLevel(6)
```

Reply:
The comment means now the varlink method does not set the log level to LOG_NULL.
It does not mean it refuses an null argument.

Reply:
I think it depends on the perspective. I can pass `null` and the API happily accepts that. I'm not saying I'm opposed that (after all I mentioned polkit with its half-implemented LogLevel and hardcoded LogTarget as an example of why I think LogLevel can be the only place where setting loglevel to null can be actually implemented eventually). I just think that the API (if it's supposed to be generic) should stick to one meaning. `null` resetting the log level in some places and `null` setting the log target to `null` in some other places seems weird to me. Either way it hasn't been publicly released yet so it can be changed (or whatever).

Reply:
In that sense, I mostly agree. At least our implementation should stick `null` -> `LOG_INFO` (or any behavior is OK if it is consistent to all services) unless there is a strong reason.